components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-provider.ts

```diff
@@ -68,31 +74,84 @@ export class GitpodUserStorageProvider implements FileSystemProviderWithFileRead
-        const server = service.server;
         const content = this.encodingService.decode(BinaryBuffer.wrap(value));
         // TODO server could accept binary instead
         const uri = this.toBackwardCompatibleUri(resource);
-        await server.updateUserStorageResource({ uri, content });
-        this.updateStat(resource, value);
+        await this.cachedWrite({ uri, content });
+        this.updateStat(resource, value, true);
         // TODO server could fire a notification that some resournce was changed instead
         this.onDidChangeFileEmitter.fire([{
             type: FileChangeType.UPDATED,
             resource
         }]);
     }
 
+    // To avoid having badly written extensions executing too much requests and putting pressure on other components we
+    // cache writes and reads here.
+    protected readonly contentCache = new Map<string, string>();
+    protected updatedSinceLastWrite = new Set<string>();
+    protected async cachedWrite(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.add(resource.uri);
+        
+        console.count(`user storage write to: ${resource.uri}`);
+    }
+
+    protected async cachedRead(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+
+        const content = await this.getServer().getUserStorageResource(req);
+        this.contentCache.set(req.uri, content);
+        return content;
+    }
+
+    protected startPeriodicCacheWriter(interval: number) {
+        const performWrites = async () => {
+            const writeStart = Date.now();
+
+            const updatedUris = this.updatedSinceLastWrite;
+            this.updatedSinceLastWrite = new Set<string>();
+
+            const server = this.getServer();
+            for (const uri of updatedUris) {
+                let gotDisconnectedDuringUpdate = false;
+                let listener = server.onDidCloseConnection(_ => {
+                    gotDisconnectedDuringUpdate = true;
+                    listener.dispose();
+                });
+
+                try {
+                    const content = this.contentCache.get(uri)!;
+                    await server.updateUserStorageResource({ uri, content });
+                } catch (err) {
+                    if (gotDisconnectedDuringUpdate) {
+                        this.updatedSinceLastWrite.add(uri);
+                    } else {
+                        console.error(`failed to update '${uri}' resource`, err);
+                    }
+                } finally {
+                    listener.dispose();
+                }
+            }
+
+            const waitFor = (interval * 1000) - (Date.now() - writeStart);
+            setTimeout(performWrites, waitFor > 0 ? waitFor : 0)
```

Comment:
Is this complexity really worth it? I'd say just passing on `interval` as is would be good.

Reply:
> I'd say just passing on interval as is would be good.

I started out with that, but while discussing with the others someone (I think @akosyakov ) feared that after a re-connect it takes 10s to save the last version. This calculation is meant to fix that.